### PR TITLE
Add `lazy_init` method to `Strategy` class

### LINKED ISSUE #7650: Make lazy initialization in plugins more robust
## 🚀 Feature

We currently set some attributes like sync_batch_norm, num_nodes, devices etc. lazily in the accelerator connector, so the user does not have to provide them at instantiation. 

Example:

Trainer(gpus=4, plugins=DDPPlugin(find_unused_parameters=True))  # plugin may require gpus, num_nodes etc.

# AcceleratorConnector does this:
training_type_plugin.num_nodes = ...

This is fragile. Some attributes may have to be recomputed based on the order in which others are set.

### Pitch

Provide one single lazy init method that takes all arguments required. The plugin is responsible for making sure dependencies are resolved in one place: 

class DDPPlugin(...):

    def lazy_init(self, **kwargs):
         self.num_nodes = kwargs.get("num_nodes")
         self.num_processes = ...

### PR DESCRIPTION
**Pull request by edward-io**
## What does this PR do?

open questions:

- should we pass in any kwargs, or be explicit in what we should accept for lazy initialization?

Part of #7650

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by edward-io** [comment]
Planning on splitting this PR into two: one introducing the new method, another implementing it in strategies.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -721,25 +722,16 @@ def _validate_precision_choice(self) -> None:
...
+        self.strategy.lazy_init(
+            accelerator=self.accelerator,
+            precision_plugin=self._precision_plugin,
+            checkpoint_io=self.checkpoint_io,
+            cluster_environment=self.cluster_environment,
+            parallel_devices=self._parallel_devices,
+            num_nodes=self._num_nodes_flag,
+            layer_sync=self._layer_sync,
+        )
+        self._parallel_devices = self.strategy.parallel_devices
```
Comment (four4fish) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:734]: we need actually implement the lazy_init() then change accelerator_connector. 
I think we can have one PR introduce lazy_init in strategies (it will be great to have them in one PR, easier to compare), the second one enable the call to lazy_init()

**Comment by four4fish** [comment]
I think this will not make to 1.6 release and this won't be user facing change. How about we target at 1.6.x

**Comment by kaushikb11** [comment]
@edward-io This is great. Any updates on this?

**Comment by Borda** [comment]
@awaelchli, mind checking this regarding the ongoing Lite refactoring? :otter:

**Comment by awaelchli** [comment]
@Borda I have continued the work in https://github.com/Lightning-AI/lightning/pull/13399, but have hit minor blockers. IMO it is low priority atm, but can be investigated after our Lite efforts.
